xarray/core/alignment.py

```diff
@@ -358,7 +366,15 @@ def var_indexers(var, indexers):
             args = (var.attrs, var.encoding)
         else:
             args = ()
-        reindexed[dim] = IndexVariable((dim,), indexers[dim], *args)
+
+        idx_var = IndexVariable((dim,), indexers[dim], *args)
+
+        # GH1434
+        # ensures that dtype of numpy structured arrays is preserved
+        if len(args) and is_composite_dtype(var.dtype) \
+                and idx_var.dtype != var.dtype:
+            idx_var.data = idx_var.data.astype(var.dtype)
```

Comment:
I still don't understand why this change is necessary. Can you give an example of what this would fix?